### PR TITLE
Amazon Q Feature - Use language level from project and module settings

### PR DESCRIPTION
**Pull request by damntrecky**
## Types of changes

- [ x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Description
The goal of this PR is to allow project validation to read the project and module level "language level" to help determine what Java version is being used.

IntelliJ references Maven `<properties>` and sets the appropriate "language level" at the project and module level.

```
  <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <java.version>8</java.version>
    <maven.compiler.release>8</maven.compiler.release>
    <maven.compiler.source>1.8</maven.compiler.source>
    <maven.compiler.target>1.8</maven.compiler.target>
  </properties>
```

Will set "language level" to 8 in your module.

### Impact
This will impact users who cannot transform because their JDK/SDK version is set to JDK 17 or higher. We will allow users who are compiling Java 8/11 projects from JDK 17+.

Previously if users had sdk set to JDK17 they automatically failed clientside validation and the project could not be submitted. We believe this will be a substantial change to help transform projects

## Checklist
- [x ] My code follows the code style of this project
- [x ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtils.kt [hunk 1]
```diff
@@ -64,18 +73,6 @@ fun Project.getSupportedBuildFilesWithSupportedJdk(
     }
 }
 
-fun Project.getSupportedBuildModulesPath(supportedBuildFileNames: List<String>): List<String> {
```
Comment (damntrecky) [inline plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtils.kt:67]: Comment: Unused code deleted

File: plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.codemodernizer.utils
```
Comment (damntrecky) [inline plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt:4]: TODO: Adding unit tests here as well. Currently in progress.

File: plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformModuleUtilsTest.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (damntrecky) [inline plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformModuleUtilsTest.kt:1]: TODO: Tests currently failing. Fixing locally right now.

File: plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformModuleUtils.kt [hunk 4]
```diff
@@ -7,14 +7,32 @@ import com.intellij.openapi.module.Module
...
+ * The default value should be set to the project settings SDK and language level, so if the parent SDK is set to
+ * Java 17 and the language level is set to default. The value spit out will be JDK_17. If the parent language
+ * level is set to JDK_1_8 then the default will be JDK_1_8 for the module. You can override all this at the
+ * module "source" setting
+ */
+fun Module.tryGetJdkLanguageLevelJdk(): JavaSdkVersion? {
+    val moduleRootManager = ModuleRootManager.getInstance(this)
+    val languageLevelModuleExtension = moduleRootManager.getModuleExtension(LanguageLevelModuleExtensionImpl::class.java)
+    val languageLevel = languageLevelModuleExtension?.languageLevel
+    return JavaSdkVersion.fromLanguageLevel(languageLevel!!)
```
Comment (tincheng) [inline plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformModuleUtils.kt:37]: Do we need to update any existing UX messages/documentation that QCT will try to infer sdk version from language level? Also ask customers to ensure their project can build properly when using older language level.

Comment (damntrecky) [reply plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformModuleUtils.kt]: Yes, that's another discussion. We don't see this as a one way door as the IDE already trys to set language level.

We mostly need messaging updates around this. In the short term we want to get this out and see if it solves some customer issues that are currently happening.

**Review by rli — APPROVED** [review]
same comment as why we need the change in `package-lock.json`

**Comment by rli** [comment]
i don't think 232 understands that JDK21 exists

```
e: file:///codebuild/output/src3285644532/src/github.com/aws/aws-toolkit-jetbrains/plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt:11:48 Unresolved reference: getMockJdk21
e: file:///codebuild/output/src3285644532/src/github.com/aws/aws-toolkit-jetbrains/plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt:28:19 Unresolved reference: getMockJdk21
```

File: plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Before
+    override fun setup() {
+        super.setup()
+        Mockito.mockStatic(LanguageLevelProjectExtension::class.java)
+
+        sdkMock = getMockJdk21()
+        languageLevelProjectExtensionMock = Mockito.mock(LanguageLevelProjectExtension::class.java)
+
+        Mockito.`when`(LanguageLevelProjectExtension.getInstance(project)).thenReturn(languageLevelProjectExtensionMock)
```
Comment (rli) [inline plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt:31]: future prefer `whenever` from mockito-kotlin

File: plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.codemodernizer.utils
+
+import com.intellij.openapi.projectRoots.JavaSdkVersion
+import com.intellij.openapi.projectRoots.Sdk
+import com.intellij.openapi.roots.LanguageLevelProjectExtension
+import com.intellij.openapi.roots.ProjectRootManager
+import com.intellij.pom.java.LanguageLevel
+import com.intellij.testFramework.IdeaTestUtil.getMockJdk21
```
Comment (rli) [inline plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt:11]: resolve failure by making a new `tst-233+` folder and extracting test to that directory

Comment (damntrecky) [reply plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt:11]: I don't understand the workaround. I need to make an entire new folder for all new tests? Just this one time?

Comment (damntrecky) [reply plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt:11]: My local build is working

File: plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt (hunk 5 above)
Comment (damntrecky) [reply plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt:31]: Trying to update all instances right now to use the `import org.mockito.kotlin` important for `whenever` , `mock` and `doReturn`.

File: plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt (hunk 6 above)
Comment (rli) [reply plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt:11]: The active IDE version in your local project is declared in `gradle.properties`
https://github.com/aws/aws-toolkit-jetbrains/blob/main/gradle.properties#L11

Occasionally we run into situations where implementation needs to be specific to a major version range. In this case your test references code that only exists in 233+, which can be merged back into the standard `src`/`tst` folders when we drop 232.

Bottom of https://github.com/aws/aws-toolkit-jetbrains/blob/main/REPOLAYOUT.md#subprojects has a brief description

File: plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt (hunk 5 above)
Comment (damntrecky) [reply plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt:31]: Looks like all but `mockStatic` exist in `mockito-kotlin`. Replaced all instances and have tests passing again.

File: plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt (hunk 6 above)
Comment (damntrecky) [reply plugins/amazonq/codetransform/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformProjectUtilsTest.kt:11]: Thank you for the detailed documentation and explanation. I will let me team know about this as well during standup.

Updated. Cheers!

**Review by rli — APPROVED** [review]

**Review by dhasani23 — APPROVED** [review]

**Pull request merged**